### PR TITLE
refactor: remove `Validation` from functions in `Weeder2`

### PR DESCRIPTION
**Pull request by jaschdoc**
Depends on https://github.com/flix/flix/pull/8961

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 1]
```diff
@@ -73,56 +72,60 @@ object Weeder2 {
...
+      case None => List.empty
+    }
   }
 
   private def visitUse(tree: Tree)(implicit sctx: SharedContext): Validation[List[UseOrImport], CompilationMessage] = {
     expect(tree, TreeKind.UsesOrImports.Use)
     val maybeUseMany = tryPick(TreeKind.UsesOrImports.UseMany, tree)
-    flatMapN(pickQName(tree))(qname => {
-      if (qname.namespace.idents.isEmpty && maybeUseMany.isEmpty) {
+    mapN(pickQName(tree)) { qname =>
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:92]: removed flatmap here

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 2]
```diff
@@ -73,56 +72,60 @@ object Weeder2 {
...
-        case None => List(UseOrImport.Use(qname, qname.ident, qname.loc))
-        // case: use many
-        case Some(uses) => uses
+        List.empty
+      } else {
+        val nname = Name.NName(qname.namespace.idents :+ qname.ident, qname.loc)
+        val optUses = maybeUseMany.map(visitUseMany(_, nname))
+        optUses match {
+          // case: use many, if uses is empty the parser will have reported an error
+          case Some(uses) => uses
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:105]: merged duplicate `Some` case

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 3]
```diff
@@ -73,56 +72,60 @@ object Weeder2 {
...
+  private def visitUseMany(tree: Tree, namespace: Name.NName)(implicit sctx: SharedContext): List[UseOrImport] = {
     expect(tree, TreeKind.UsesOrImports.UseMany)
-    val identUses = traverse(pickAll(TreeKind.Ident, tree))(visitUseIdent(_, namespace))
-    val aliasedUses = traverse(pickAll(TreeKind.UsesOrImports.Alias, tree))(tree => visitUseAlias(tree, namespace))
-    mapN(identUses, aliasedUses) {
-      (identUses, aliasedUses) => (identUses ++ aliasedUses).sortBy(_.loc)
-    }
+    val identUses = pickAll(TreeKind.Ident, tree).map(visitUseIdent(_, namespace))
+    val aliasedUses = pickAll(TreeKind.UsesOrImports.Alias, tree).map(visitUseAlias(_, namespace))
+    (identUses ++ aliasedUses).sortBy(_.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:118]: removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 4]
```diff
@@ -73,56 +72,60 @@ object Weeder2 {
...
   }
 
-  private def visitUseIdent(tree: Tree, namespace: Name.NName)(implicit sctx: SharedContext): Validation[UseOrImport.Use, CompilationMessage] = {
-    mapN(tokenToIdent(tree)) {
-      ident => UseOrImport.Use(Name.QName(namespace, ident, tree.loc), ident, ident.loc)
-    }
+  private def visitUseIdent(tree: Tree, namespace: Name.NName)(implicit sctx: SharedContext): UseOrImport.Use = {
+    val ident = tokenToIdent(tree)
+    UseOrImport.Use(Name.QName(namespace, ident, tree.loc), ident, ident.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:123]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 5]
```diff
@@ -131,61 +134,59 @@ object Weeder2 {
...
 
       // recover from missing alias by using ident
       case ident :: _ =>
         val error = Malformed(NamedTokenSet.Alias, SyntacticContext.Use, hint = Some(s"Give an alias after ${TokenKind.ArrowThickR.display}."), loc = tree.loc)
         sctx.errors.add(error)
         val qname = Name.QName(namespace, ident, tree.loc)
-        Validation.success(UseOrImport.Use(qname, ident, ident.loc))
+        UseOrImport.Use(qname, ident, ident.loc)
+
       case _ => throw InternalCompilerException("Parser passed malformed use with alias", tree.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:146]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 6]
```diff
@@ -131,61 +134,59 @@ object Weeder2 {
...
-        case Some(Nil) => List.empty
-        // case: import one, use the java name
-        case None =>
-          val ident = Name.Ident(jname.fqn.lastOption.getOrElse(""), jname.loc)
-          List(UseOrImport.Import(jname, ident, tree.loc))
-        // case: import many
-        case Some(imports) => imports
-      }
-    })
+    mapN(JvmOp.pickJavaName(tree)) {
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:152]: Removed `flatMapN`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 7]
```diff
@@ -131,61 +134,59 @@ object Weeder2 {
...
+      jname =>
+        val maybeImportMany = tryPick(TreeKind.UsesOrImports.ImportMany, tree).map(visitImportMany(_, jname.fqn))
+        maybeImportMany match {
+          // case: import many, if the list is empty parser has reported an error
+          case Some(imports) => imports
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:157]: merged duplicate `Some` case

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 8]
```diff
@@ -131,61 +134,59 @@ object Weeder2 {
...
+  private def visitImportMany(tree: Tree, namespace: Seq[String])(implicit sctx: SharedContext): List[UseOrImport.Import] = {
     expect(tree, TreeKind.UsesOrImports.ImportMany)
-    val identImports = traverse(pickAll(TreeKind.Ident, tree))(visitImportIdent(_, namespace))
-    val aliasedImports = traverse(pickAll(TreeKind.UsesOrImports.Alias, tree))(tree => visitImportAlias(tree, namespace))
-    mapN(identImports, aliasedImports) {
-      (identImports, aliasedImports) => (identImports ++ aliasedImports).sortBy(_.loc)
-    }
+    val identImports = pickAll(TreeKind.Ident, tree).map(visitImportIdent(_, namespace))
+    val aliasedImports = pickAll(TreeKind.UsesOrImports.Alias, tree).map(visitImportAlias(_, namespace))
+    (identImports ++ aliasedImports).sortBy(_.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:171]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 9]
```diff
@@ -131,61 +134,59 @@ object Weeder2 {
...
+        UseOrImport.Import(jname, alias, tree.loc)
       // recover from missing alias by using ident
       case ident :: _ =>
         val error = Malformed(NamedTokenSet.Alias, SyntacticContext.Import, hint = Some(s"Give an alias after ${TokenKind.ArrowThickR.display}."), loc = tree.loc)
         sctx.errors.add(error)
-        Validation.success(UseOrImport.Import(Name.JavaName(Seq(ident.name), tree.loc), ident, ident.loc))
+        UseOrImport.Import(Name.JavaName(Seq(ident.name), tree.loc), ident, ident.loc)
       case _ => throw InternalCompilerException("Parser passed malformed use with alias", tree.loc)
     }
   }
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:192]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 10]
```diff
@@ -930,20 +933,17 @@ object Weeder2 {
...
-      }
+      val idents = pickAll(TreeKind.Ident, tree).map(tokenToIdent)
+      assert(idents.nonEmpty) // Require at least one ident
+      val first = idents.head
+      val ident = idents.last
+      val nnameIdents = idents.dropRight(1)
+      val loc = SourceLocation(isReal = true, first.loc.sp1, ident.loc.sp2)
+      val nname = Name.NName(nnameIdents, loc)
+      val qname = Name.QName(nname, ident, loc)
+      Expr.Ambiguous(qname, qname.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:946]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 11]
```diff
@@ -1012,13 +1012,11 @@ object Weeder2 {
...
+    private def visitHoleExpr(tree: Tree)(implicit sctx: SharedContext): Expr = {
       expect(tree, TreeKind.Expr.Hole)
-      mapN(tryPickNameIdent(tree)) {
-        ident =>
-          val strippedIdent = ident.map(id => Name.Ident(id.name.stripPrefix("?"), id.loc))
-          Expr.Hole(strippedIdent, tree.loc)
-      }
+      val ident = tryPickNameIdent(tree)
+      val strippedIdent = ident.map(id => Name.Ident(id.name.stripPrefix("?"), id.loc))
+      Expr.Hole(strippedIdent, tree.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:1019]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 12]
```diff
@@ -1737,7 +1735,7 @@ object Weeder2 {
     private def visitWithoutExpr(tree: Tree)(implicit sctx: SharedContext): Validation[Expr, CompilationMessage] = {
       expect(tree, TreeKind.Expr.Without)
       val effectSet = pick(TreeKind.Type.EffectSet, tree)
-      val effects = flatMapN(effectSet)(effectSetTree => traverse(pickAll(TreeKind.QName, effectSetTree))(visitQName))
+      val effects = mapN(effectSet)(effectSetTree => pickAll(TreeKind.QName, effectSetTree).map(visitQName))
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:1738]: Removed `flatMapN`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 13]
```diff
@@ -2009,24 +2007,25 @@ object Weeder2 {
     private def visitFixpointInjectExpr(tree: Tree)(implicit sctx: SharedContext): Validation[Expr, CompilationMessage] = {
       expect(tree, TreeKind.Expr.FixpointInject)
       val expressions = pickAll(TreeKind.Expr.Expr, tree)
-      val idents = pickAll(TreeKind.Ident, tree)
-      flatMapN(traverse(expressions)(visitExpr), traverse(idents)(tokenToIdent)) {
-        case (exprs, idents) if exprs.length != idents.length =>
+      val idents = pickAll(TreeKind.Ident, tree).map(tokenToIdent)
+      mapN(traverse(expressions)(visitExpr)) {
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2011]: Removed `flatMapN`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 14]
```diff
@@ -2704,24 +2701,26 @@ object Weeder2 {
...
         case TreeKind.ErrorTree(_) => Validation.success(Type.Error(tree.loc))
         case kind => throw InternalCompilerException(s"Parser passed unknown type '$kind'", tree.loc)
       }
     }
 
-    private def visitNameType(tree: Tree)(implicit sctx: SharedContext): Validation[Type, CompilationMessage] = {
-      mapN(visitQName(tree))(Type.Ambiguous(_, tree.loc))
+    private def visitNameType(tree: Tree)(implicit sctx: SharedContext): Type.Ambiguous = {
+      val qname = visitQName(tree)
+      Type.Ambiguous(qname, tree.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2715]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 15]
```diff
@@ -2704,24 +2701,26 @@ object Weeder2 {
...
-      mapN(tokenToIdent(tree)) {
-        case ident if ident.isWild => Type.Var(ident, tree.loc)
-        case ident => Type.Ambiguous(Name.QName(Name.RootNS, ident, ident.loc), tree.loc)
-      }
+    private def visitIdentType(tree: Tree)(implicit sctx: SharedContext): Type = {
+      val ident = tokenToIdent(tree)
+      if (ident.isWild)
+        Type.Var(ident, tree.loc)
+      else
+        Type.Ambiguous(Name.QName(Name.RootNS, ident, ident.loc), tree.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2723]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 16]
```diff
@@ -2760,30 +2759,25 @@ object Weeder2 {
     }
 
     private def visitSchemaRowType(parentTree: Tree)(implicit sctx: SharedContext): Validation[Type, CompilationMessage] = {
-      val maybeRest = tryPick(TreeKind.Ident, parentTree)
-      flatMapN(traverseOpt(maybeRest)(tokenToIdent)) {
-        maybeRest =>
-          val rest = maybeRest match {
-            case None => WeededAst.Type.SchemaRowEmpty(parentTree.loc)
-            case Some(name) => WeededAst.Type.Var(name, name.loc)
+      val rest = tryPick(TreeKind.Ident, parentTree).map(tokenToIdent) match {
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2762]: Removed `flatMapN`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 17]
```diff
@@ -2760,30 +2759,25 @@ object Weeder2 {
     }
 
     private def visitSchemaRowType(parentTree: Tree)(implicit sctx: SharedContext): Validation[Type, CompilationMessage] = {
-      val maybeRest = tryPick(TreeKind.Ident, parentTree)
-      flatMapN(traverseOpt(maybeRest)(tokenToIdent)) {
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2764]: Removed `flatMapN`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 18]
```diff
@@ -2884,10 +2878,10 @@ object Weeder2 {
       }
     }
 
-    private def visitCaseSetType(tree: Tree)(implicit sctx: SharedContext): Validation[Type, CompilationMessage] = {
+    private def visitCaseSetType(tree: Tree)(implicit sctx: SharedContext): Type.CaseSet = {
       expect(tree, TreeKind.Type.CaseSet)
-      val cases = traverse(pickAll(TreeKind.QName, tree))(visitQName)
-      mapN(cases)(Type.CaseSet(_, tree.loc))
+      val cases = pickAll(TreeKind.QName, tree).map(visitQName)
+      Type.CaseSet(cases, tree.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2884]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 19]
```diff
@@ -2910,11 +2904,10 @@ object Weeder2 {
...
     }
 
-    private def visitVariableType(tree: Tree)(implicit sctx: SharedContext): Validation[Type.Var, CompilationMessage] = {
+    private def visitVariableType(tree: Tree)(implicit sctx: SharedContext): Type.Var = {
       expect(tree, TreeKind.Type.Variable)
-      mapN(tokenToIdent(tree)) {
-        ident => Type.Var(ident, tree.loc)
-      }
+      val ident = tokenToIdent(tree)
+      Type.Var(ident, tree.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:2910]: Removed `Validation`

**Review by magnus-madsen — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 20]
```diff
@@ -3079,11 +3071,12 @@ object Weeder2 {
       val signature = pickSignature(tree)
       val ascription = pickAscription(tree)
       val ident = tryPickNameIdent(tree)
-      mapN(fqn, signature, ascription, ident) {
-        case (fqn, signature, (tpe, eff), ident) => if (isStatic)
-          WeededAst.JvmOp.StaticMethod(fqn, signature, tpe, eff, ident)
-        else
-          WeededAst.JvmOp.Method(fqn, signature, tpe, eff, ident)
+      mapN(fqn, signature, ascription) {
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:3074]: moved `ident` out and simplified body

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 21]
```diff
@@ -3135,30 +3128,27 @@ object Weeder2 {
   }
 
   private def pickQName(tree: Tree)(implicit sctx: SharedContext): Validation[Name.QName, CompilationMessage] = {
-    flatMapN(pick(TreeKind.QName, tree))(visitQName)
+    mapN(pick(TreeKind.QName, tree))(visitQName)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:3131]: Removed `flatMapN`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 22]
```diff
@@ -3135,30 +3128,27 @@ object Weeder2 {
...
-        Name.QName(nname, ident, loc)
-    }
+    val idents = pickAll(TreeKind.Ident, tree).map(tokenToIdent)
+    assert(idents.nonEmpty) // We require at least one element to construct a qname
+    val first = idents.head
+    val ident = idents.last
+    val nnameIdents = idents.dropRight(1)
+    val loc = SourceLocation(isReal = true, first.loc.sp1, ident.loc.sp2)
+    val nname = Name.NName(nnameIdents, loc)
+    Name.QName(nname, ident, loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:3143]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 23]
```diff
@@ -3135,30 +3128,27 @@ object Weeder2 {
...
   }
 
   private def pickNameIdent(tree: Tree)(implicit sctx: SharedContext): Validation[Name.Ident, CompilationMessage] = {
-    flatMapN(pick(TreeKind.Ident, tree))(tokenToIdent)
+    mapN(pick(TreeKind.Ident, tree))(tokenToIdent)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:3147]: Removed `flatMapN`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 24]
```diff
@@ -3135,30 +3128,27 @@ object Weeder2 {
...
   }
 
-  private def tryPickNameIdent(tree: Tree)(implicit sctx: SharedContext): Validation[Option[Name.Ident], CompilationMessage] = {
-    traverseOpt(tryPick(TreeKind.Ident, tree))(tokenToIdent)
+  private def tryPickNameIdent(tree: Tree)(implicit sctx: SharedContext): Option[Name.Ident] = {
+    tryPick(TreeKind.Ident, tree).map(tokenToIdent)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:3151]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 25]
```diff
@@ -3183,22 +3173,22 @@ object Weeder2 {
...
       case Some(token@Token(_, _, _, _, sp1, sp2)) =>
-        Validation.success(Name.Ident(token.text, SourceLocation(isReal = true, sp1, sp2)))
+        Name.Ident(token.text, SourceLocation(isReal = true, sp1, sp2))
       // If child is an ErrorTree, that means the parse already reported and error.
       // We can avoid double reporting by returning a success here.
       // Doing it this way is most resilient, but phases down the line might have trouble with this sort of thing.
       case Some(t: Tree) if t.kind.isInstanceOf[TreeKind.ErrorTree] =>
         val name = text(tree).mkString("")
-        Validation.success(Name.Ident(name, tree.loc))
+        Name.Ident(name, tree.loc)
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:3185]: Removed `Validation`

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 26]
```diff
@@ -3183,22 +3173,22 @@ object Weeder2 {
     * Only use this if the structure of tree is well-known.
     * IE. Calling on a tree of kind [[TreeKind.Ident]] is fine, but if the kind is not known avoid using [[tokenToIdent]].
     */
-  private def tokenToIdent(tree: Tree)(implicit sctx: SharedContext): Validation[Name.Ident, CompilationMessage] = {
+  private def tokenToIdent(tree: Tree)(implicit sctx: SharedContext): Name.Ident = {
     tree.children.headOption match {
       case Some(token@Token(_, _, _, _, sp1, sp2)) =>
-        Validation.success(Name.Ident(token.text, SourceLocation(isReal = true, sp1, sp2)))
+        Name.Ident(token.text, SourceLocation(isReal = true, sp1, sp2))
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:3179]: Removed `Validation`

**Pull request merged**